### ISSUE TITLE
Discussion about ProgressBar and metrics display

### ISSUE THREAD

**Issue by milesial** [pytorch/ignite#296]
Hello, I would like to discuss the feature in `contrib.handlers.ProgressBar` that displays metrics on each `Events.ITERATION_COMPLETED` as a postfix of the bar. For some context, please read https://github.com/pytorch/ignite/pull/295.

I think the `ProgressBar` and this feature are essential to the package, as it provides a concise way to have a nice training and evaluation experience (replacing the old way of spamming numbers in the console), and is appealing to beginners.

Here is some of the issues I found.

### Problem
On the doc, the given example is
```python
pbar = ProgressBar()
pbar.attach(trainer, ['loss'])
```

The `metrics` parameter (here `['loss']`) works with `Metric`s that have been attached to the engine, **AND that rewired `self.computed` (which updates the `state.metrics` dict of the engine) on `Events.ITERATION_COMPLETED` instead of `Events.EPOCH_COMPLETED`**. The second part is unintuitive, and not documented. The current document suggests that this might work with all metrics, but this raises a `KeyError` when the dict has not been updated before the end of the first iteration.
Currently, this freature only works with `RunningAverage`.

The problem has been raised by @vfdev-5 in https://github.com/pytorch/ignite/pull/256, but I did not see the discussion go anywhere, I would like for it to continue.

### Proposed solutions

1. The documentation should be updated to reflect the fact that this feature will not work on all of the built-in metrics but `RunningAverage`.

2. The error message for the `KeyError` is `metrics not found in engine.state.metrics` and could be modified to hint for what the issue might be.

3. a Maybe the `Metric` class hierarchy should be updated, to distinguish metrics that update the engine state every epoch, and those that update it every iteration. The `ProgressBar` could then only accept the latter one. 

3. b Or maybe all of the current built-in metrics could update the state of the engine at each iteration (on `self.update` instead of on `self.completed`. This will add computation at each epoch. The progress bar would then work on all metrics.

### Problem
Have a look at the doc example again :
```python
pbar = ProgressBar()
pbar.attach(trainer, ['loss'])
```
This suggests that we can print the training loss, that updates at each iteration.

However, we can't. When we are using a trainer created with `engine.create_supervised_trainer`, the `process_function` returns the loss for the processed batch. I think this is the most intuitive way to do it even for custom training engines.
The `Metric`s that can be attached are computed from the output of this function (`y_pred, y` for an evaluator for example), which may pass through a transform function. You see where I am going: the output of the trainer is already the loss itself, we do not need any further computation.

If we want to print the training loss, we have to attach a `Metric` that does nothing but moves the already computed loss into the `state.metrics` dict of the engine at every iteration. See https://github.com/pytorch/ignite/pull/295 for the `Identity` metric. This is not intuitive, we should be able to display the training loss out of the box.

### Proposed solutions

1. Built-in Identity metric, see https://github.com/pytorch/ignite/pull/295 for the code.

2. Modify `ProgressBar` to allow the display in the postfix the value of `engine.state.output`
    The `attach` method could then be 
    ```python
    attach(self, engine, metric_names=None, display_output=False)
    ```
    with the output looking like 
    `Epoch 1: [59/689]   9%|▊                 , output=3.54e-02 [00:04<00:50]`
3. Modify `ProgressBar` to allow the display in the postfix the value of any custom variables in the `engine.state` object, including the output.

---
Thanks for reading, sorry for the long issue and the english mistakes.

**Comment by miguelvr** [comment]
Hey @milesial, thanks for your input.

> Built-in Identity metric, see #295 for the code.

I agree that we should have a wrapper for the engine output to be saved as a metric, this would be the easiest and cleaner solution.

> Modify ProgressBar to allow the display in the postfix the value of engine.state.output

I think this is off the table because different engines can have different outputs, so it wouldn't generalize.

> Modify ProgressBar to allow the display in the postfix the value of any custom variables in the engine.state object, including the output.

this could also be interesting, but it may be tricky to generalize to all the possible cases. We had a discussion when I first implemented the `ProgressBar` about handling the output as well, but the code was quickly getting messy and confusing.

@vfdev-5 what's your opinion on these?

**Comment by milesial** [comment]
> I think this is off the table because different engines can have different outputs, so it wouldn't generalize.

I agree, one solution could be to ask for a function `output_to_str`, so the user is responsible for formatting the output. If no such function is passed, we can assume the returned variable is a float, string, ..., or has a `__repr__` method.

**Comment by vfdev-5** [comment]
Concerning the issue, current implementation of `ProgressBar` actually works with other metrics if they are already computed. I agree that we need to improve the documentation on this.

Concerning the output and custom variables. What do you think about a solution as it is done for [`RunningAverage`](https://github.com/pytorch/ignite/blob/master/ignite/metrics/running_average.py#L9) ?

So a new API of `ProgressBar` can look like : 
```python
pbar = ProgressBar()
# current behaviour
pbar.attach(trainer, metric_names=['some_running_metric', ], output_transform=None)

# display transformed output
pbar.attach(trainer, metric_names=None, output_transform=lambda x: x)

# display custom variable
pbar.attach(trainer, metric_names=None, output_transform=None, custom_variables=['custom_var', ])

# progress bar only: default values
pbar.attach(trainer, metric_names=None, output_transform=None, custom_variables=None)
```

This is not a perfect solution if user has multiple outputs to display with a custom messages etc. But in this situation I wonder if user should not use `log_message` as all shown progress is not logged persistently and even removed when epoch ends.

**Comment by miguelvr** [comment]
I can make an option to persist the bar...

**Comment by milesial** [comment]
About the persistency, I was thinking about displaying the training loss for each batch, of course the loss of the last batch will be overwritten by the loss of the new batch. This allows us to look at the bar progression and see the evolution of the loss without getting spammed. At the end of the epoch, a handler can be used to print persistently the total average loss of the epoch.

**Comment by vfdev-5** [comment]
> About the persistency, I was thinking about displaying the  training loss for each batch, of course the loss of the last batch will  be overwritten by the loss of the new batch.

Actually, why you do not want to use a running average with loss at each batch ? As all discussion we started from this...

**Comment by milesial** [comment]
I want to see the non-smoothed value of the current loss (this is why I proposed alpha=0 in https://github.com/pytorch/ignite/pull/295)

**Comment by milesial** [comment]
So in the API, `output_transform` should return a printable object that the `ProgressBar` can use ? What if the returned object is already printable : a float, string, ... ? passing `lambda x: x` seems unnecessary.

**Comment by milesial** [comment]
Okay, it would be nice to be able to pass `output_transform=True`, and it would treat it as `output_transform=lambda x: x`. But then the name `output_transform` is strange, we could rename it to `display_output`.

So we can pass the following into the `display_output` parameter :
- `None` or `False`: do not display output
- a callable : pass through it before displaying
- `True` : display without going through a callable (or applying `lambda x: x`, same effect)

In my mind, in the case of `create_supervised_trainer`, it would be cleaner to simply be able to :
```python
pbar.attach(trainer, display_output=True)
```
With the ability to pass callable to process different outputs, as you proposed.

**Comment by vfdev-5** [comment]
We have already introduced `output_transform` before in Metrics and I think we should keep same API for similar stuff here and not introduce something different instead. And keep thing simple.

-    `None`: do not display output
-    a callable : pass through it before displaying.

**Comment by milesial** [comment]
Okay, I can live with that.
But for the `custom_variables` we assume they are all printable ? A list of tuples (var_name, output_transform) seems complicated.

**Comment by vfdev-5** [comment]
Yes, we can explicitly detail in the doc that progress bar can display only printable variables.

**Comment by miguelvr** [comment]
> But for the `custom_variables` we assume they are all printable ? A list of tuples (var_name, output_transform) seems complicated.

you would actually have to transform it into a dictionary `{name: value}` for it to make sense

> we could rename it to `display_output`

This doesn't generalize, you are only thinking about the `create_supervised_trainer` case. If you look at the DCGAN example you'll see the `step` function output it is actually a dictionary, for example.
